File: src/harkerrobolib/subsystems/HSArm.java
package harkerrobolib.subsystems;

import com.ctre.phoenix.motorcontrol.ControlMode;
import com.ctre.phoenix.motorcontrol.DemandType;

import edu.wpi.first.wpilibj2.command.SubsystemBase;
import harkerrobolib.wrappers.HSMotorController;

/**
 * Represents a general Arm subsystem with a single motor controller.
 *
 * @param talon the arm's Talon
 *
 * @author Finn Frankis
 * @author Angela Jia
 * @version 10/31/18
 */
public abstract class HSArm<Motor extends HSMotorController> extends SubsystemBase {

	private Motor talon;
    private final double feedForwardGrav;
    
    public enum ArmDirection{
        UP(1), DOWN(-1);
    	private int sign;
    	private ArmDirection(int sign) {
    		this.sign = sign;
    	}
    }
	
	public HSArm(Motor talon, double feedForwardGrav)
	{
		this.talon = talon;
		this.feedForwardGrav = feedForwardGrav;
    }

    public void setCurrentLimits(int peakTime, int peakCurrent, int contCurrent) {
        talon.configPeakCurrentDuration(peakTime);
        talon.configPeakCurrentLimit(peakCurrent);
        talon.configContinuousCurrentLimit(contCurrent);
        talon.enableCurrentLimit(true);
    }

    public void armMotionPercentOutput(double output, ArmDirection direction) {
        armMotionPercentOutput(direction.sign * output);
    }

    public void armMotionPercentOutput(double output) {
        talon.set(ControlMode.PercentOutput, output, DemandType.ArbitraryFeedForward, feedForwardGrav);
    }

    public double getTalonCurrent() {
    	return talon.getOutputCurrent();
    }
    
    public Motor getTalon() {
        return talon;
    }
    
    public double getFeedForwardGrav() {
        return feedForwardGrav;
    }
}

File: src/harkerrobolib/subsystems/HSArm.kt
package harkerrobolib.subsystems

import com.ctre.phoenix.motorcontrol.ControlMode
import com.ctre.phoenix.motorcontrol.DemandType
import edu.wpi.first.wpilibj.command.Subsystem
import harkerrobolib.wrappers.HSTalon

/**
 * Represents a general Arm subsystem with a single motor controller.
 *
 * @author Finn Frankis
 * @author Angela Jia
 * @version 10/31/18
 */
abstract class HSArm(val talon : HSTalon, val feedForwardGrav : Double = 0.0 ) : Subsystem(){

    enum class ArmDirection (val sign : Int) {
        UP(1), DOWN(-1);

        operator fun times (other : ArmDirection) =  sign * other.sign
        operator fun times (other : Double) = sign * other
    }

    fun setCurrentLimits( peakTime : Int, peakCurrent : Int, contCurrent : Int, timeout : Int = talon.defaultTimeout ) {
        with (talon) {
            configPeakCurrentDuration(peakTime, timeout)
            configPeakCurrentLimit(peakCurrent, timeout)
            configContinuousCurrentLimit(contCurrent, timeout)
            enableCurrentLimit(true)
        }
    }

    fun armMotionPercentOutput(output: Double, direction: ArmDirection) {
        armMotionPercentOutput(direction * output)
    }

    fun armMotionPercentOutput (output: Double) {
        talon.set(ControlMode.PercentOutput, output, DemandType.ArbitraryFeedForward, feedForwardGrav)
    }

    fun getTalonCurrent() = talon.outputCurrent
}
